core/src/main/scala/org/apache/spark/ui/viz/VisualizationListener.scala

```diff
@@ -0,0 +1,104 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.ui.viz
+
+import scala.collection.mutable
+import scala.xml.{Node, Unparsed}
+
+import org.apache.spark.SparkConf
+import org.apache.spark.scheduler._
+import org.apache.spark.ui.SparkUI
+
+/**
+ * A SparkListener that constructs RDD DAG visualization for the UI.
+ */
+private[ui] class VisualizationListener(conf: SparkConf) extends SparkListener {
+  private val jobIdToStageIds = new mutable.HashMap[Int, Seq[Int]]
+  private val stageIdToGraph = new mutable.HashMap[Int, VizGraph]
+  private val stageIds = new mutable.ArrayBuffer[Int]
+
+  // How many jobs or stages to retain graph metadata for
+  private val retainedStages =
+    conf.getInt("spark.ui.retainedStages", SparkUI.DEFAULT_RETAINED_STAGES)
+
+  /** Construct a "Show visualization" DOM element that expands into a visualization for a stage. */
+  def showVizElementForStage(stageId: Int): Seq[Node] = {
+    showVizElement(getVizGraphForStage(stageId).toSeq, forJob = false)
+  }
+
+  /** Construct a "Show visualization" DOM element that expands into a visualization for a job. */
+  def showVizElementForJob(jobId: Int): Seq[Node] = {
+    showVizElement(getVizGraphsForJob(jobId), forJob = true)
+  }
+
+  /** Construct a "Show visualization" DOM element that expands into a visualization on the UI. */
+  private def showVizElement(graphs: Seq[VizGraph], forJob: Boolean): Seq[Node] = {
+    <div>
```

Comment:
The only thing is that this code is used in both the job page and the stage page. If we put them in those respective files then there will be some duplicate code.